optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,329 @@
+                x = trial.suggest_float("x", -1, 1)
+                y = trial.suggest_int("y", -1, 1)
+                return x ** 2 + y
+
+
+            sampler = optuna.samplers.QMCSampler()
+            study = optuna.create_study(sampler=sampler)
+            study.optimize(objective, n_trials=8)
+
+    """
+
+    def __init__(
+        self,
+        *,
+        qmc_type: str = "halton",
+        scramble: bool = False,  # default is False for simplicity in distributed environment.
+        seed: Optional[int] = None,
+        independent_sampler: Optional[BaseSampler] = None,
+        warn_asyncronous_seeding: bool = True,
+        warn_independent_sampling: bool = True,
+    ) -> None:
+
+        self._scramble = scramble
+        self._seed = seed or np.random.PCG64().random_raw()
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._initial_search_space: Optional[Dict[str, BaseDistribution]] = None
+        self._warn_independent_sampling = warn_independent_sampling
+
+        if qmc_type in ("halton", "sobol"):
+            self._qmc_type = qmc_type
+        else:
+            message = (
+                f"The `qmc_type`, {qmc_type}, is not a valid. "
+                'It must be one of "halton" and "sobol".'
+            )
+            raise ValueError(message)
+
+        if (seed is None) and scramble and warn_asyncronous_seeding:
+            # Sobol/Halton sequences without scrambling do not use seed.
+            self._log_asyncronous_seeding()
+
+    def reseed_rng(self) -> None:
+
+        # We must not reseed the `self._seed` like below. Otherwise, workers will have different
+        # seed under parallel execution because `self.reseed_rng()` is called when starting each
+        # parallel executor.
+        # >>> self._seed = np.random.MT19937().random_raw()
+
+        self._independent_sampler.reseed_rng()
+
+    def infer_relative_search_space(
+        self, study: Study, trial: FrozenTrial
+    ) -> Dict[str, BaseDistribution]:
+
+        if self._initial_search_space is not None:
+            return self._initial_search_space
+
+        past_trials = study.get_trials(deepcopy=False, states=_SUGGESTED_STATES)
+        # The initial trial is sampled by the independent sampler.
+        if len(past_trials) == 0:
+            return {}
+        # If an initial trial was already made,
+        # construct search_space of this sampler from the initial trial.
+        first_trial = min(past_trials, key=lambda t: t.number)
+        self._initial_search_space = self._infer_initial_search_space(first_trial)
+        return self._initial_search_space
+
+    def _infer_initial_search_space(self, trial: FrozenTrial) -> Dict[str, BaseDistribution]:
+
+        # TODO(kstoneriv3): Replace `OrderedDict` to `Dict` after
+        # the support for Python 3.6 is stopped.
+        search_space: OrderedDict[str, BaseDistribution] = OrderedDict()
```

Comment:
Since the QMCSampler is not supported in python 3.6, can we replace this `OrderedDict` to `Dict`?